test/e2e_awssharedvpc/aws_shared_vpc_test.go

```diff
@@ -0,0 +1,307 @@
+		fmt.Printf("Failed to init kube client: %v\n", err)
+		os.Exit(1)
+	}
+
+	openshiftCI = common.IsOpenShift()
+	err, platformType = common.GetPlatformType(openshiftCI, kubeClient)
+	if err != nil {
+		fmt.Printf("Failed to determine platform type: %v\n", err)
+		os.Exit(1)
+	}
+
+	if common.SkipProvider(platformType) {
+		fmt.Printf("Skipping e2e test for the provider %q!\n", platformType)
+		os.Exit(0)
+	}
+
+	// Only run this test if the DNS config contains the privateZoneIAMRole which indicates it's a "Shared VPC" cluster.
+	// Note: Only AWS supports privateZoneIAMRole.
+	dnsConfig := configv1.DNS{}
+	err = kubeClient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, &dnsConfig)
+	if err != nil {
+		fmt.Printf("Failed to get dns 'cluster': %v\n", err)
+		os.Exit(1)
+	}
+	if dnsConfig.Spec.Platform.AWS == nil || dnsConfig.Spec.Platform.AWS.PrivateZoneIAMRole == "" {
+		fmt.Printf("Test skipped on non-shared-VPC cluster\n")
+		os.Exit(0)
+	}
+	roleARN = dnsConfig.Spec.Platform.AWS.PrivateZoneIAMRole
+	hostedZoneID = dnsConfig.Spec.PrivateZone.ID
+	hostedZoneDomain = dnsConfig.Spec.BaseDomain
+
+	err, r53ClientAssumeRole = initRoute53ClientAssumeRole(openshiftCI, kubeClient, roleARN)
+
+	if err := common.EnsureOperandResources(kubeClient); err != nil {
+		fmt.Printf("Failed to ensure operand resources: %v\n", err)
+	}
+
+	exitStatus := m.Run()
+	os.Exit(exitStatus)
+}
+
+// TestExternalDNSAssumeRole tests the assumeRole functionality in which you can specify a Role ARN to use another
+// account's hosted zone for creating DNS records. Only AWS is supported.
+func TestExternalDNSAssumeRole(t *testing.T) {
+	t.Log("Ensuring test namespace")
+	err := kubeClient.Create(context.TODO(), &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		t.Fatalf("Failed to ensure namespace %s: %v", testNamespace, err)
+	}
+
+	// Create an External object that uses the role ARN in the dns config to create DNS records in the private DNS
+	// zone in another AWS account route 53.
+	t.Log("Creating ExternalDNS object that assumes role our of private zone in another account's route 53")
+	extDNS := buildExternalDNSAssumeRole(testExtDNSName, hostedZoneID, hostedZoneDomain, roleARN)
+	if err := kubeClient.Create(context.TODO(), &extDNS); err != nil {
+		t.Fatalf("Failed to create external DNS %q: %v", testExtDNSName, err)
+	}
+	defer func() {
+		_ = kubeClient.Delete(context.TODO(), &extDNS)
+	}()
+
+	// Create a service of type LoadBalancer with the annotation targeted by the ExternalDNS resource.
+	t.Log("Creating source service")
+	service := common.DefaultService(testServiceName, testNamespace)
+	if err := kubeClient.Create(context.TODO(), service); err != nil {
+		t.Fatalf("Failed to create test service %s/%s: %v", testNamespace, testServiceName, err)
+	}
+	defer func() {
+		_ = kubeClient.Delete(context.TODO(), service)
```

Comment:
Hmm, yes that is true. Though, for existing tests, we delete the hosted zone: https://github.com/openshift/external-dns-operator/blob/d04af909927c7c9f47d308ae14e9744b7842c238/test/e2e/operator_test.go#L137

So we can be sure they eventually get deleted. 

For this new Shared VPC e2e test, we just delete service, but the private hosted zone doesn't get deleted (because we didn't create it).

What do you think about adding a function in the `defer` for deleting the service to wait for ExternalDNS to delete the records (and error if they weren't cleaned up)? Right now, it may be a race condition situation (1. delete service, then immediately 2. delete ExternalDNS object).

Reply:
Update: Added logic in the new E2E test that deletes the service and then waits for the DNS record to get deleted.